option --TBsearch

robertnurnberg opened the pull request:
After discussions with noob on sf-discord, by default we no longer continue searches into 7men TBs.

I believe this needs some testing. In particular, I am not sure that 7men positions with castling rights are dealt with correctly. All moves in such positions are now assigned a value of 1cp from `queryall` (not 0, to differentiate them from true EGTB draws), and we only terminate the search of a line once this has been resolved to draw or win/loss.

Sample output for a TB position with cr for `python cdbsearch.py --epd "r3k3/8/4K3/8/8/8/8/4Q3 w q -"`:
```
Root position:  r3k3/8/4K3/8/8/8/8/4Q3 w q -
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-05T19:51:12.920550
Search at depth  1
  cdb PV len:  0
  score     :  -1
  PV        :  e6f6 e8f8
  PV len    :  2
  max ply   :  1
  queryall  :  24
  bf        :  24.00
  inflightR :  9.62
  inflightQ :  11.54
  chessdbq  :  24
  enqueued  :  24
  unscored  :  0
  date      :  2023-06-05T19:51:14.961980
  total time:  0:00:02.04
  cdb time  :  85
  URL       :  https://chessdb.cn/queryc_en/?r3k3/8/4K3/8/8/8/8/4Q3_w_q_-_moves_e6f6_e8f8

Search at depth  2
  cdb PV len:  0
  score     :  24992
  PV        :  e6d6 e8f8 e1f1 EGTB
  PV len    :  3
  max ply   :  1
  queryall  :  295
  bf        :  17.18
  inflightR :  9.52
  inflightQ :  114.04
  chessdbq  :  271
  enqueued  :  24
  unscored  :  0
  date      :  2023-06-05T19:51:21.092712
  total time:  0:00:08.17
  cdb time  :  30
  URL       :  https://chessdb.cn/queryc_en/?r3k3/8/4K3/8/8/8/8/4Q3_w_q_-_moves_e6d6_e8f8_e1f1

Search at depth  3
  cdb PV len:  0
  score     :  24992
  PV        :  e6d6 EGTB
  PV len    :  1
  max ply   :  1
  queryall  :  296
  bf        :  6.66
  inflightR :  9.52
  inflightQ :  114.04
  chessdbq  :  271
  enqueued  :  24
  unscored  :  0
  date      :  2023-06-05T19:51:21.366349
  total time:  0:00:08.44
  cdb time  :  31
  URL       :  https://chessdb.cn/queryc_en/?r3k3/8/4K3/8/8/8/8/4Q3_w_q_-_moves_e6d6
```

Searches for EGTB positions terminate immediately, e.g. `python cdbsearch.py --epd "3qk3/8/8/8/3Pp3/8/8/3QK3 b - d3"` gives
```
Root position:  3qk3/8/8/8/3Pp3/8/8/3QK3 b - d3
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-05T19:51:52.210946
Search at depth  1
  cdb PV len:  1
  score     :  0
  PV        :  d8h4 draw
  PV len    :  1
  max ply   :  0
  queryall  :  3
  bf        :  3.00
  inflightR :  0.67
  inflightQ :  2.00
  chessdbq  :  3
  enqueued  :  0
  unscored  :  0
  date      :  2023-06-05T19:51:52.848107
  total time:  0:00:00.63
  cdb time  :  212
  URL       :  https://chessdb.cn/queryc_en/?3qk3/8/8/8/3Pp3/8/8/3QK3_b_-_d3_moves_d8h4
```

An example for an 8men position with best move a capture into 7men is this: `python cdbsearch.py --epd "4k3/ppp2p2/8/8/4p3/8/4Q3/4K3 w - - 0 1" --depthLimit 1`
```
Root position:  4k3/ppp2p2/8/8/4p3/8/4Q3/4K3 w - - 0 1
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-05T20:32:25.336233
Search at depth  1
  cdb PV len:  1
  score     :  24996
  PV        :  e2e4 e8f8 e4b7 EGTB
  PV len    :  3
  max ply   :  2
  queryall  :  7
  bf        :  7.00
  inflightR :  0.57
  inflightQ :  2.14
  chessdbq  :  7
  enqueued  :  0
  unscored  :  0
  date      :  2023-06-05T20:32:26.752795
  total time:  0:00:01.41
  cdb time  :  202
  URL       :  https://chessdb.cn/queryc_en/?4k3/ppp2p2/8/8/4p3/8/4Q3/4K3_w_-_-_0_1_moves_e2e4_e8f8_e4b7
```

Here I do not understand why the PV shows 3 moves. Given the code, I would expect `e2e4 e8f8 EGTB`. Maybe you see the reason? (Edit: The bug has since been fixed, the original PR miscounted the pieces.)

PS: All the `max ply` outputs are from before my last commit, where I corrected the `-1` bit.

robertnurnberg commented on cdbsearch.py:
```diff
@@ -523,15 +557,19 @@ async def cdbsearch(
                 )
         print(f"{pv[-1]}\n  PV len    :  {pvlen}")
-        print(f"  max ply   :  {len(chessdb.semaphoreTree) - 1}")
+        print(f"  max ply   :  {len(chessdb.semaphoreTree)}")
```
I believe the -1 was incorrect. Agreed?

vondele replied:
Not so sure, I've added it there because we have `self.semaphoreTree[ply]`, in that case `len(semaphreTree)` is ply + 1 ? Am I missing something ?

robertnurnberg replied:
I am going to untangle this PR into the separate pieces. It sort of grew out of control as I was making the TBsearch edits in my branch. There I will explain my reasoning.

robertnurnberg commented on cdbsearch.py:
```diff
@@ -361,6 +368,24 @@ async def search(self, board, depth):
+            if abs(bestscore) >= CDB_SPECIAL:
+                return bestscore - (1 if bestscore >= 0 else -1), [bestmove, "EGTB"]
+            # hence bestscore is 1 or -1, which means we have a TB position with castling rights: continue the search
+            assert (
```
This assert we can remove after sufficient testing, I guess?

robertnurnberg commented:
This cleaned-up PR contains just the stuff related to TBsearch. I also fixed a bug in the piece counting. So with this update, the examples from the original post (plus one more) read as follows:

- `python cdbsearch.py --epd "r3k3/8/4K3/8/8/8/8/4Q3 w q -"` (4men position w/ cr: search stops completely once win is found)
```
Root position:  r3k3/8/4K3/8/8/8/8/4Q3 w q -
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-06T09:31:35.891256
Search at depth  1
  cdb PV len:  0
  score     :  -1
  PV        :  e6f6 e8f8
  PV len    :  2
  max ply   :  2
  queryall  :  24
  bf        :  24.00
  inflightR :  9.62
  inflightQ :  11.54
  chessdbq  :  24
  enqueued  :  24
  unscored  :  0
  date      :  2023-06-06T09:31:38.382564
  total time:  0:00:02.49
  cdb time  :  103
  URL       :  https://chessdb.cn/queryc_en/?r3k3/8/4K3/8/8/8/8/4Q3_w_q_-_moves_e6f6_e8f8

Search at depth  2
  cdb PV len:  0
  score     :  24992
  PV        :  e6d6 e8f8 e1f1 EGTB
  PV len    :  3
  max ply   :  2
  queryall  :  295
  bf        :  17.18
  inflightR :  8.73
  inflightQ :  114.04
  chessdbq  :  271
  enqueued  :  24
  unscored  :  0
  date      :  2023-06-06T09:31:49.883835
  total time:  0:00:13.99
  cdb time  :  51
  URL       :  https://chessdb.cn/queryc_en/?r3k3/8/4K3/8/8/8/8/4Q3_w_q_-_moves_e6d6_e8f8_e1f1

Search at depth  3
  cdb PV len:  0
  score     :  24992
  PV        :  e6d6 EGTB
  PV len    :  1
  max ply   :  2
  queryall  :  296
  bf        :  6.66
  inflightR :  8.73
  inflightQ :  114.04
  chessdbq  :  271
  enqueued  :  24
  unscored  :  0
  date      :  2023-06-06T09:31:50.189428
  total time:  0:00:14.29
  cdb time  :  52
  URL       :  https://chessdb.cn/queryc_en/?r3k3/8/4K3/8/8/8/8/4Q3_w_q_-_moves_e6d6
```

- `python cdbsearch.py --epd "3qk3/8/8/8/3Pp3/8/8/3QK3 b - d3"` (6men w/ ep: search stops immediately)
```
Root position:  3qk3/8/8/8/3Pp3/8/8/3QK3 b - d3
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-06T09:36:14.118716
Search at depth  1
  cdb PV len:  1
  score     :  0
  PV        :  d8h4 draw
  PV len    :  1
  max ply   :  0
  queryall  :  3
  bf        :  3.00
  inflightR :  0.67
  inflightQ :  2.00
  chessdbq  :  3
  enqueued  :  0
  unscored  :  0
  date      :  2023-06-06T09:36:14.796057
  total time:  0:00:00.67
  cdb time  :  225
  URL       :  https://chessdb.cn/queryc_en/?3qk3/8/8/8/3Pp3/8/8/3QK3_b_-_d3_moves_d8h4
  ```

- `python cdbsearch.py --epd "4k3/ppp2p2/8/8/4p3/8/4Q3/4K3 w - - 0 1" --depthLimit 1` (8men w/ bestmove capture: search would continue indefinitely, as root is not in 7men)
```
Root position:  4k3/ppp2p2/8/8/4p3/8/4Q3/4K3 w - - 0 1
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-06T09:38:13.964234
Search at depth  1
  cdb PV len:  1
  score     :  24998
  PV        :  e2e4 e8d7 EGTB
  PV len    :  2
  max ply   :  1
  queryall  :  4
  bf        :  4.00
  inflightR :  0.67
  inflightQ :  2.00
  chessdbq  :  3
  enqueued  :  0
  unscored  :  0
  date      :  2023-06-06T09:38:14.817708
  total time:  0:00:00.85
  cdb time  :  284
  URL       :  https://chessdb.cn/queryc_en/?4k3/ppp2p2/8/8/4p3/8/4Q3/4K3_w_-_-_0_1_moves_e2e4_e8d7
  ```
  
-  `python cdbsearch.py --epd "4k3/ppp2p2/8/8/4p3/8/8/3QK3 w - - 0 1" --depthLimit 1` (8men w/ bestmove no capture: longer PV before "EGTB")
```
Root position:  4k3/ppp2p2/8/8/4p3/8/8/3QK3 w - - 0 1
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-06-06T09:42:38.916875
Search at depth  1
  cdb PV len:  3
  score     :  24994
  PV        :  d1a4 e8f8 a4e4 b7b6 EGTB
  PV len    :  4
  max ply   :  3
  queryall  :  10
  bf        :  10.00
  inflightR :  1.33
  inflightQ :  2.89
  chessdbq  :  9
  enqueued  :  1
  unscored  :  0
  date      :  2023-06-06T09:42:40.648618
  total time:  0:00:01.73
  cdb time  :  192
  URL       :  https://chessdb.cn/queryc_en/?4k3/ppp2p2/8/8/4p3/8/8/3QK3_w_-_-_0_1_moves_d1a4_e8f8_a4e4_b7b6
  ```

robertnurnberg commented on cdbsearch.py:
```diff
@@ -360,6 +368,24 @@ async def search(self, board, depth):
+                key=lambda t: t[1],
+            )
+            if bestscore == 0:
+                return 0, [bestmove, "draw"]
```
So at the moment we return "draw" for a TB draw, but "EGTB" for win/loss (user can check eval, I did not want to introduce too many terminal cases). Arguably, we could also return "EGTB" for TB draws, user can see from 0 eval that it is a draw. What do you think?

vondele replied:
I think we could use EGTB for both TB win and draw, since these are 'proven' results. That is different from a typical draw eval.

robertnurnberg commented:
Rebased on current main. I think ready for review (but my comments on code should be resolved before merging.)

robertnurnberg commented:
I attach the output of the command `python3 cdbbulksearch.py ../matetrack/matetrack.epd --bulkConcurrency 48 --concurrency 48 --depthLimit 2 --user rob >& cdbbulk_notbsearch.log &` for my `egtb` branch. I believe the output shows that this works as expected.

[cdbbulk_notbsearch.log](https://github.com/vondele/cdbexplore/files/11738699/cdbbulk_notbsearch.log)

vondele commented on cdbsearch.py:
```diff
@@ -373,6 +377,24 @@ async def search(self, board, depth):
+        # stop search if we are in EGTB and know the result
+        if (
+            not self.TBsearch
+            and sum(p in "pnbrqk" for p in board.epd().lower().split()[0]) <= CDB_EGTB
```
I think this can be done without searching and string manipulation in the epd. For example this is counting code I have elsewhere:
```
                    knights = bin(board.knights).count("1")
                    bishops = bin(board.bishops).count("1")
                    rooks = bin(board.rooks).count("1")
                    queens = bin(board.queens).count("1")
                    pawns = bin(board.pawns).count("1")

                    matcountkey = (
                        9 * queens + 5 * rooks + 3 * knights + 3 * bishops + pawns
                    )
```

robertnurnberg replied:
Sure. But I do not think there is a big difference in efficiency here, and my code is 1 line and quite pythonic. :)

If you agree with the logic of this PR, we could leave non-functional changes for a follow-up PR?

vondele replied:
yes, can be done in a follow up.

robertnurnberg commented:
I changed "draw" to "EGTB" for TB draws and removed the assert. I am re-running the test from https://github.com/vondele/cdbexplore/pull/56#issuecomment-1589919615 and will post logs here. Then this could be merged, IMO.
